lnrpc/invoicerpc/invoice.proto

```diff
@@ -0,0 +1,385 @@
+syntax = "proto3";
+
+package invoicerpc;
+
+
+/**
+InvoiceService has the suffix Service to prevent name clash with the Invoice
+message.
+*/
+service InvoiceService {
+
+    /** lncli: `addinvoice`
+    AddInvoice attempts to add a new invoice to the invoice database. Any
+    duplicated invoices are rejected, therefore all invoices *must* have a
+    unique payment preimage.
+    */
+    rpc AddInvoice (AddInvoiceRequest) returns (AddInvoiceResponse) {}
+
+    /** lncli: `listinvoices`
+    ListInvoices returns a list of all the invoices currently stored within the
+    database. Any active debug invoices are ignored. It has full support for
+    paginated responses, allowing users to query for specific invoices through
+    their invoice_number. This can be done by using either the first_index_offset or
+    last_index_offset fields included in the response as the index_offset of the
+    next request. The reversed flag is set by default in order to paginate
+    backwards. If you wish to paginate forwards, you must explicitly set the
+    flag to false. If none of the parameters are specified, then the last 100
+    invoices will be returned.
+    */
+    rpc ListInvoices (ListInvoiceRequest) returns (ListInvoiceResponse) {}
+
+    /** lncli: `lookupinvoice`
+    LookupInvoice attempts to look up an invoice according to its payment hash.
+    The passed payment hash *must* be exactly 32 bytes, if not, an error is
+    returned.
+    */
+    rpc LookupInvoice (PaymentHash) returns (Invoice) {}
+
+    /**
+    SubscribeInvoices returns a uni-directional stream (server -> client) for
+    notifying the client of invoice updates. The caller can optionally specify
+    a last_update_index to catch up with missed updates.
+    */
+    rpc SubscribeInvoices (InvoiceSubscription) returns (stream Invoice) {}
```

Comment:
Instead of this global subscribe function (is it scalable?), we could also go for a SubscribeInvoice with a specific invoice number. Isn't it that applications (except for some kind of global node monitoring dashboard), know their own invoices?

Reply:
Then we also wouldn't need all those indices. Just start the stream by sending back the current invoice state and keep sending updates as they happen. It will be a never ending stream in case we decide to keep accepting htlcs even though the invoice is settled. But when the client isn't interested anymore, it can just close the stream.

This `SubscribeInvoice` can even be combined with `AddInvoice`. Create an idempotent `AddInvoice`.

Reply:
Agreed! It would certainly be useful to also be able to filter for a specific invoice (or maybe even a list?)

Reply:
How do you feel about an idempotent, streaming `AddInvoice`? It can still be used like the current `AddInvoice`. Just close the stream after the first (creation) update is received.

Reply:
Another question that I have with the existing global `SubscribeInvoices` is when to store those last indices? Updating a persistent store after every event doesn't sound good. But using a timer is also not so clean.

Reply:
`SubscribeInvoices` is nice for a global monitoring dashboard, but I suspect that most other applications know their own invoices and are much better served with a per invoice subscription.